ksonnet-gen/ksonnet/marshaller.go

```diff
@@ -0,0 +1,61 @@
+package ksonnet
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+)
+
+// marshaller abstracts the task of writing out indented text to a
+// buffer. Different components can call `indent` and `dedent` as
+// appropriate to specify how indentation needs to change, rather than
+// to keep track of the current indentation.
+//
+// For example, if one component is responsible for writing an array,
+// and an element in that array is a function, the component
+// responsible for the array need only know to call `indent` after the
+// '[' character and `dedent` before the ']' character, while the
+// routine responsible for writing out the function can handle its own
+// indentation independently.
+type marshaller struct {
+	depth  int
+	prefix string
+	lines  []string
+	buffer *bytes.Buffer
+}
+
+func newMarshaller() *marshaller {
+	var buffer bytes.Buffer
+	return &marshaller{
+		depth:  0,
+		prefix: "",
+		lines:  []string{},
+		buffer: &buffer,
+	}
+}
+
+func (m *marshaller) bufferLine(text string) {
```

Comment:
So, this is these comments about `io.Writer` are the most interesting part of the review for me, but the one I have the last intuition for what the best thing to do is. So per your suggestion, I will write a `TODO` to revisit it, adopt a bunch of your suggestions about this stuff, check that in. Then, we can revisit when we get a chance to sync on Monday perhaps.

I ended up designing this to minimize stuff you need to remember at the callsite because my estimation was that it didn't dramatically hurt testing to not use `io.Writer`, while it did potentially hurt correctness to have to remember to check an error and emit a newline when using something like `fmt.Fprintf`.

The design we ultimately landed on is partially informed by Rob Pike's little post on [errors as values](https://blog.golang.org/errors-are-values), in which he proposes that if you don't want to check `if err != nil` after every operation, you can write a little struct like:

```
type errWriter struct {
    w   io.Writer
    err error
}
func (ew *errWriter) write(buf []byte) {
    if ew.err != nil {
        return
    }
    _, ew.err = ew.w.Write(buf)
}
```

This lets you chain together a bunch of `write` operations and then get an error only when you try to get the result.

You could pass an `io.Writer` to the `indentWriter`, but I kind of like that `bufferLine` semantically captures that you're emitting a line that's indented, rather than having a `write` that indents on newlines. Particularly I like this because it forces the user to think about each line they're emitting, and whether it should be indented.

I don't know, thoughts? It's hard to tell how off the beaten path this really is.